### DISCUSSION TITLE
AsyncFileStreamer example is unclear

### DISCUSSION THREAD

**Discussion by paulharris** [uNetworking/uWebSockets#1780] [Q&A]
Hi,
Getting started with uWebSockets is made more difficult because the AsyncFileStreamer.h (in particular) file seems incomplete and written by someone who doesn't know how uWebSockets works.

Example: 
```
                // todo: I don't really know what this is supposed to mean?
                return false;
```
https://github.com/uNetworking/uWebSockets/blob/388e2b3364ffcd4bc2c57a73166ecfde8fc1875d/examples/helpers/AsyncFileStreamer.h#L78

This in particular is in `streamFile()`, which is part of the lambda created and passed to `res->onWritable()`.
The documentation states that `You MUST return true for success, false for failure.` but in this case, the lambda seems to be simply scheduling a future attempt to send some more of the file in the future.

There is more confusion in the example... can we see a complete example written with best practices please?

**Comment by uNetworkingAB** [comment]
That example is unfinished

**Comment by e3dio** [comment]
>  // todo: I don't really know what this is supposed to mean?

> onWritable().
The documentation states that You MUST return true for success, false for failure

The return value of `onWritable` is not relevant, has no effect since this commit https://github.com/uNetworking/uWebSockets/commit/5528e6f5fec2ffbd8e684e5d7d966eb0ae2bd5d5 Could probably update the docs to clarify that

**Comment by e3dio** [comment]
If it helps here is example in JS:
```js
import { open } from 'node:fs/promises';

const streamFile = async (res, path, headers) => {
	res.onAborted(() => res.aborted = true);
	const fh = await open(path), size = (await fh.stat()).size;
	if (res.aborted) return fh.close();
	if (headers) res.cork(() => { for (const h in headers) res.writeHeader(h, headers[h]); });
	for await (const chunk of fh.createReadStream()) {
		if (res.aborted) break;
		let ok, done;
		const lastOffset = res.getWriteOffset();
		res.cork(() => [ ok, done ] = res.tryEnd(chunk, size));
		if (!ok) done = await new Promise(resolve => {
			res.onAborted(() => resolve(true));
			res.onWritable(offset => {
				res.cork(() => [ ok, done ] = res.tryEnd(chunk.subarray(offset - lastOffset), size));
				if (ok) resolve(done); 
			});
		});
		if (done) break;
	}
	return fh.close();
};

app.get('/', r => streamFile(r, path))
```

**Comment by paulharris** [comment]
The comment left in that code (at the time of the commit: If write was
never called, the developer should still return true so that we may drain.

Yes it would be nice if documentation were updated.

**Comment by paulharris** [comment]
Thanks, would much prefer a C++ example that was maintained with the
code...

The existing example doesn't even do the cork stuff?

**Reply by porsager** [reply]
Great opportunity for a PR when you figure it out 😉

**Reply by paulharris** [reply]
Because the best person to rewrite the example is someone with almost zero experience with uWebSockets?

**Reply by uNetworkingAB** [reply]
You can look at VideoStreamer.js example in uWS.js repo. Ask ChatGPT to rewrite it in C++ because uWS.js and uWS have the same interfaces

**Reply by paulharris** [reply]
I'm not much of a ChatGPT user.  I gave it a try and it gave me generic C++ with traditional sockets.  A second time it did it with boost/beast.
Never mind, I guess the focus for this library is to support the Node/JS world.  I'll look for another library.
Good work though!  Looks fantastic.

**Reply by uNetworkingAB** [reply]
sounds good
